crates/rattler_shell_helpers/src/lib.rs

```diff
@@ -0,0 +1,418 @@
+/// Collect all environment variables that are set in a conda environment.
+/// The environment variables are collected from the `state` file and the `env_vars.d` directory in the given prefix
+/// and are returned as a ordered map.
+///
+/// # Arguments
+///
+/// * `prefix` - The path to the root of the conda environment
+///
+/// # Returns
+///
+/// A map of environment variables
+///
+/// # Errors
+///
+/// If the `state` file or the `env_vars.d` directory cannot be read, an error is returned.
+fn collect_env_vars(prefix: &Path) -> Result<IndexMap<String, String>, ActivationError> {
+    let state_file = prefix.join("conda-meta/state");
+    let pkg_env_var_dir = prefix.join("etc/conda/env_vars.d");
+    let mut env_vars = IndexMap::new();
+
+    if pkg_env_var_dir.exists() {
+        let env_var_files = pkg_env_var_dir.read_dir()?;
+
+        let mut env_var_files = env_var_files
+            .into_iter()
+            .filter_map(|r| r.ok())
+            .map(|e| e.path())
+            .filter(|path| path.is_file())
+            .collect::<Vec<_>>();
+
+        // sort env var files to get a deterministic order
+        env_var_files.sort();
+
+        let env_var_json_files = env_var_files
+            .iter()
+            .map(|path| {
+                fs::read_to_string(path)?
+                    .parse::<serde_json::Value>()
+                    .map_err(|e| ActivationError::InvalidEnvVarFileJson(e, path.to_path_buf()))
+            })
+            .collect::<Result<Vec<serde_json::Value>, ActivationError>>()?;
+
+        for (env_var_json, env_var_file) in env_var_json_files.iter().zip(env_var_files.iter()) {
+            let env_var_json = env_var_json.as_object().ok_or_else(|| {
+                ActivationError::InvalidEnvVarFileJsonNoObject {
+                    file: pkg_env_var_dir.to_path_buf(),
+                }
+            })?;
+
+            for (key, value) in env_var_json {
+                if let Some(value) = value.as_str() {
+                    env_vars.insert(key.to_string(), value.to_string());
+                } else {
+                    println!(
+                        "WARNING: environment variable {key} has no string value (path: {env_var_file:?})");
+                }
+            }
+        }
+    }
+
+    if state_file.exists() {
+        let state_json = fs::read_to_string(&state_file)?;
+
+        // load json but preserve the order of dicts - for this we use the serde preserve_order feature
+        let state_json: serde_json::Value = serde_json::from_str(&state_json)
+            .map_err(|e| ActivationError::InvalidEnvVarFileJson(e, state_file.to_path_buf()))?;
+
+        let state_env_vars = state_json["env_vars"].as_object().ok_or_else(|| {
+            ActivationError::InvalidEnvVarFileStateFile {
+                file: state_file.to_path_buf(),
+            }
+        })?;
+
+        for (key, value) in state_env_vars {
+            if state_env_vars.contains_key(key) {
+                println!(
```

Comment:
Can you use `tracing` instead of `println`? That way the user of the library is able to configure how they want to log things.